controllers/common/metrics.go

```diff
@@ -0,0 +1,100 @@
+package common
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	namespace = "instance_manager"
+)
+
+type MetricsCollector struct {
+	prometheus.Collector
+
+	successCounter   *prometheus.CounterVec
+	failureCounter   *prometheus.CounterVec
+	throttleCounter  *prometheus.CounterVec
+	statusGauge      *prometheus.GaugeVec
+	lastUpgradeGauge *prometheus.GaugeVec
+}
+
+func NewMetricsCollector() *MetricsCollector {
+	return &MetricsCollector{
+		successCounter: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: namespace,
+				Name:      "reconcile_success_total",
+				Help:      `total successful reconciles`,
+			},
+			[]string{"instancegroup"},
+		),
+		failureCounter: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: namespace,
+				Name:      "reconcile_fail_total",
+				Help:      `total failed reconciles`,
+			},
+			[]string{"instancegroup", "reason"},
+		),
+		throttleCounter: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: namespace,
+				Name:      "aws_api_throttle_total",
+				Help:      "number of aws API calls throttles",
+			},
+			[]string{"service", "operation"},
+		),
+		statusGauge: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Namespace: namespace,
+				Name:      "instance_group_status",
+				Help:      "number of instance groups and their status",
+			},
+			[]string{"instancegroup", "status"},
+		),
+	}
+}
+
+func (c MetricsCollector) Collect(ch chan<- prometheus.Metric) {
+	c.successCounter.Collect(ch)
+	c.failureCounter.Collect(ch)
+	c.throttleCounter.Collect(ch)
+	c.statusGauge.Collect(ch)
+}
+
+func (c MetricsCollector) Describe(ch chan<- *prometheus.Desc) {
+	c.successCounter.Describe(ch)
+	c.failureCounter.Describe(ch)
+	c.throttleCounter.Describe(ch)
+	c.statusGauge.Describe(ch)
+}
+
+func (c *MetricsCollector) SetInstanceGroup(instanceGroup, state string) {
+	externalStates := []string{"ReconcileModifying", "InitUpgrade", "Deleting", "Ready", "Error"}
```

Comment:
Are these states constants?

Reply:
Yes, in v1alpha1 - this would be a cyclic dependency if we use them though